Lecture 12: Theory of Variational Inference: Marginal Polytope, Inner and Outer Approximation (#14)
sailinglab/pgm-spring-2019: _posts/2019-02-25-lecture-12.md

---
layout: distill
title: "Lecture 12: Theory of Variational Inference: Marginal Polytope, Inner and Outer Approximation"
description: Introduction of Loopy Belief Propagation algorithm and the theory behind it and Mean-field approximation.
date: 2019-02-25

lecturers:
  - name: Eric Xing
    url: "https://www.cs.cmu.edu/~epxing/"

authors:
  - name: Donghan Yu  # author's full name
    url: "#"  # optional URL to the author's homepage
  - name: Songwei Ge
    url: "#"
  - name: Xiaofei Shi
    url: "#"
  - name: Aniketh Janardhan Reddy
    url: "https://anikethjr.github.io "
    
editors:
  - name: Maruan Al-Shedivat
    url: "#"  # optional URL to the editor's homepage

abstract: >
  An example abstract block.
---

## Recap of the message passing algorithm and its properties

> In the previous lectures, we have looked at exact inference algorithms and observed that they are not very efficient. Hence, a need for more efficient inference algorithms arises. This lecture will focus on such algorithms which are called Approximate Inference algorithms.

Reviewer: ```suggestion
In the previous lectures, we have looked at exact inference algorithms and observed that they are not very efficient. Hence, a need for more efficient inference algorithms arises. This lecture will focus on such algorithms which are called *Approximate Inference Algorithms*.
```

        </div>
    </div>
</figure>

The main idea of loopy belief propagation does is to extend belief propagation algorithm (a.k.a message passing algorithm) from tree to non-tree graphical models. Considering an undirected graphical model with pairwise and singleton potential functions, the loopy belief propagation algorithm calculates the messages and marginal probability based on the following equations:

<d-math block>
\begin{aligned}
M_{i\rightarrow j}(x_j) &\varpropto \sum_{x_i}\Psi_{ij}(x_i, x_j)\Psi_{i}(x_i) \prod_kM_{k\rightarrow i}(x_i) \\
b_i(x_i) &\varpropto \Psi_{i}(x_i) \prod_kM_k(x_k)
\end{aligned}
</d-math>

To be specific, the messages are updated and passed iteratively among nodes at the same time. Different from belief propagation algorithm where we pass the messages from the leaves to the root, the message passing here is recurrent. Another difference is that the loopy belief propagation algorithm doesn't need to pass messages only after collecting all the messages from its neighbors. Previous studies showed that directly copying the idea of belief propagation from tree to non-tree graphical model leads to two outcomes:

> * The algorithms converge and the approximated results are usually close to the real results which are calculated by brute-force marginalization of every other variables;

Reviewer: ```suggestion
* The algorithm converges and the approximated results are usually close to the results calculated by brute-force marginalization of variables.
```

> * The algorithm fails convergence at all and oscillate among multiple answers of $$b$$.

Reviewer: ```suggestion
* The algorithm fails to converge at all and oscillates between multiple answers of $$b$$.
```

### A theory behind loopy belief propagation: Bethe approximation to Gibbs free energy

It is often the case to use a distribution Q to approximate the intractable distribution P. In order to obtain a good approximation, the KL-divergence between Q and P is supposed to be reasonably small. Based on the factorized probability of the joint distribution, we can write the KL-divergence between Q and P as follows:

<d-math block>
\begin{aligned}

> KL(Q, P) &= \sum_XQ(X)\log(\frac{Q(X)}{P(X)}) \\

Reviewer: ```suggestion
KL(Q, P) &= \sum_XQ(X)\log\left(\frac{Q(X)}{P(X)}\right) \\
```

&= \sum_XQ(X)\log Q(X)-\sum_XP(X)\log P(X) \\
&= -H_Q(X)-E_Q\log P(X) \\
&= -H_Q(X)-\sum_{f_\alpha\in F}E_Q\log f_\alpha(X_\alpha) + \log Z 
\end{aligned}
</d-math>

<figure id="basic-structure" class="l-body">
    <div class="row">
        <div class="col two">
            <img src="{{ '/assets/img/notes/lecture-12/part2_3.jpg' | relative_url }}" />
        </div>
    </div>
</figure>

> We call the first two terms $$-H_Q(X)-\sum_{f_\alpha\in F}E_Q\log f_\alpha(X_\alpha)$$ the *Free Energy*. Now we consider an exmaple of tree-structured distribution shown above. Based on the chain rule and local Markov property of undirected graphical model, we can expand the joint probability to factorized probability. And according to the bayes rule, we can further write the joint probability into the following form:

Reviewer: ```suggestion
We call the first two terms $$-H_Q(X)-\sum_{f_\alpha\in F}E_Q\log f_\alpha(X_\alpha)$$ the *Free Energy*. Now we consider an exmaple of tree-structured distribution shown above. Based on the chain rule and local Markov property of undirected graphical model, we can expand the joint probability to factorized probability. According to the Bayes rule, we can further write the joint probability in the following form:
```

$$
P(X) = \frac{\prod_{\alpha} b_{\alpha}(X_{\alpha})}{\prod_{i} b_{i}^{d_i - 1}(x_i)}
$$

Then it's hard to calculate the Free energy $$F(X)$$. However, for a general
graph, we can choose approximate $$\hat{F}(P, Q)=F_{Bethe}$$, which has the
formulation:

$$
F_{Bethe} = \sum_{\alpha} \sum_{X_{\alpha}} b_{\alpha}(X_{\alpha}) \ln \frac{b_{\alpha}(X_{\alpha})}
{f_{\alpha}(X_{\alpha})} - \sum_{i} (1-d_i) \sum_{x_i}b_i(x_i) \ln b_i(x_i)
= - \langle f_{\alpha}(X_{\alpha}) \rangle - H_{Bethe}
$$

Note that this is equal to the exact Gibbs free energy when the factor graph
is a tree, but in general, $$H_{Bethe}$$ is not the same as the 
$$H$$ of a tree.

Then for the loopy graph above, we can write the $$F_{Bethe}$$ as:

$$
F_{Bethe} = \sum_{(i,j) \in E}F_{ij} - \sum_i d_i F_i
$$ 

### Constrained minimization of the Bethe free energy

 Then we want to solve the constrained minimization problem:

<d-math block>
\begin{aligned}
& \text{min}
& & F_{Bethe}(b(x_i), b(X_{\alpha})) \\
& \text{subject to}
& & \sum_{x_i} b_i(x_i) = 1, \sum_{X_{\alpha}|x_i} b_{\alpha}(X_{\alpha}) =
b_i(x_i)
\end{aligned}
</d-math>

We can write the Lagrange form as:

<d-math block>

> L = F_{Bethe} + \sum_i \gamma_i (\sum_{x_i} b_i(x_i) - 1)

Reviewer: ```suggestion
L = F_{Bethe} + \sum_i \gamma_i \left(\sum_{x_i} b_i(x_i) - 1\right)
```

+ \sum_{\alpha} \sum_{i \in N(\alpha)} \sum_{x_i} \lambda_{\alpha i}(x_i)

> (\sum_{X_{\alpha}|x_i} b_{\alpha}(X_{\alpha}) - b_i(x_i))

Reviewer: ```suggestion
\left(\sum_{X_{\alpha}|x_i} b_{\alpha}(X_{\alpha}) - b_i(x_i)\right)
```

</d-math>

Then we can have the zero-gradient solutions:

<figure id="basic-structure" class="l-body">
    <div class="row">
        <div class="col two">
            <img src="{{ '/assets/img/notes/lecture-12/solution-ydh.png' | relative_url }}" />
        </div>
    </div>
</figure>

A interesting finding is that, if we identify $$
\lambda_{\alpha i}(x_i) = \log(m_{i \rightarrow \alpha}(x_i)) = \log 
\prod_{b \in N(i) \neq \alpha} m_{b \to i}(x_i)

> $$, Then we get exactly the BP equations:

Reviewer: ```suggestion
$$, then we get exactly the BP equations:
```

<d-math block>
\begin{aligned}
b_i(x_i) & \propto f_i(x_i) \prod_{\alpha \in N(i)} m_{\alpha \to i}(x_i) \\
b_{\alpha}(X_{\alpha}) & \propto f_{\alpha}(X_{\alpha}) \prod_{i \in N(\alpha)}
\prod_{c \in N(i)|\alpha} m_{c \to i}(x_i)
\end{aligned}
</d-math>

## Theory of Variational Inferece

We have learned two families of approximate inference algorithms: Loopy 
belief propagation (sum-product) and mean-field approximation. Then in this
section we'll re-exam them from a unified point of view based on the 

> variational principle: Loop BP -- outer approximation, Mean filed --

Reviewer: ```suggestion
variational principle: Loop BP -- outer approximation, Mean Filed --
```

inner approximation.

### Variational Methods

"Variational" is a fancy name for optimization-based formulations, which 
represent the quantity of interest as the solution to an optimization 
problem. Actually many problems can be formulated in a variational way:

1. Eigenvalue problem: find the eigenvalue $$\lambda$$ of $$A$$, which means 
$$Ax=\lambda x$$ for any $$x$$. Then we have the Courant-Fischer for 
eigenvalues: $$\lambda_{\max}(A) = \max_{||x||_{2} =1} x^T A x$$.

2. Linear systems of equations: $$Ax=b, A \succ 0, x^{*} = A^{-1}b$$. The 
variational formulation can be: $$x^{*} = \text{argmin}_x (\frac{1}{2} x^T
Ax - b^Tx)$$. For large systems, we can apply conjugate gradient method 
to compute this efficiently.

### Inference Problems in Graphical Models

> Given an undirected graphical model, i.e.

Reviewer: ```suggestion
Given an undirected graphical model, i.e.,
```

$$
p(x) = \frac{1}{Z} \prod_{C \in \mathcal{C}} \psi_C(x_C),
$$

where $$\mathcal{C}$$ denotes the collection of cliques, one is interested in the inference of the marginal distributions

$$
p(x_i) = \sum_{x_j, j\neq i} p(x).
$$

### Ingredients: Exponential Families

> Definition: We say $$X$$ follows from an exponential family provided that the parametrized collection of density functions satisfies:

Reviewer: ```suggestion
**Definition:** We say $$X$$ follows from an exponential family provided that the parametrized collection of density functions satisfies:
```

$$
p(x;\theta) = \exp\left\{ \theta^T\phi(x) - A(\theta) \right\}, \qquad A(\theta)<\infty .
$$

> Moreover, $$\phi$$ is one of the sufficient statistics for $$\theta$$, see Larry Wasserman's lecture notes from 10/36-705 for more details and examples; and $$A$$ is usually known as the log partition function, which is convex and lower semi-continuous. Further, 

Reviewer: It'd be great if you could link to Larry's lecture notes here. Thanks.

<d-math block>
\begin{aligned}
A(\theta) &= \log \mathbb{E}_{\theta} \left[ \exp\{\theta^T\phi(X)\}\tight], \\
\frac{\partial A(\theta)}{\partial \theta} &= \mathbb{E}_{\theta} \left[ \phi(X) \right]. 
\end{aligned}
</d-math>

* Parametrize Graphical Models as Exponential Family
For undirected graphical model, if we assume

$$
\log \psi_C(x_C) = \log \psi(x_C;\theta_C) = \theta_C^T \phi(x_C),
$$

and let

$$
A(\theta) = \log Z(\theta)
$$

then

$$

> p(x;\theta) = \exp( \sum_{C \in \mathcal{C}} \theta_C^T \phi(x_C)- A(\theta)).

Reviewer: ```suggestion
p(x;\theta) = \exp\left( \sum_{C \in \mathcal{C}} \theta_C^T \phi(x_C)- A(\theta)\right).
```

Definition: For a function $$f$$, the convex conjugate dual, which is also known as the Legendre transform of $$f$$, is defined as

$$
f^* (\mu) = \sup_{\theta} \{ \theta^T\mu - f(\theta)\},
$$

and the convex conjugate dual is convex, no matter the original function is convex or not, and
Moreover, if $$f$$ is convex and lower semi-continuous, then 

$$
f(x) = \sup_{\mu} \{ \theta^T\mu - f^* (\mu)\}.
$$

* Application to Exponential Family

Let $$A$$ be the log partition function for the exponential family

$$
p(x;\theta) = \exp\left\{ \theta^T\phi(x) - A(\theta) \right\}, 
$$

The dual for $$A$$ is 

$$
A^* (\mu) = \sup\{ \theta^T \mu - A(\theta): A(\theta)<\infty\},
$$

and the stationarity condition is,

$$
\mu = \frac{\partial A(\theta)}{\partial \theta} = \mathbb{E}_{\theta} \left[ \phi(X) \right],
$$

we can thus represent $$\theta$$ through the mean parameter $$\mu$$.
Therefore, we have the following Legendre mapping:

$$
A^* (\mu) = \mathbb{E}_{\theta(\mu)} \left[ \log p(X;\theta(\mu)) \right] = -H(p(X;\theta(\mu))),
$$

where $$H$$ is the Boltzmann-Shannon entropy function. 

### Ingredients: Convex Polytope

Half-plane Representation: the Minkowski-Weyl Theorem

> Theorem: A non-empty convex polytope $$\mathcal{M}$$ can be characterized by a finite collection of linear inequality constraints, i.e.  

Reviewer: ```suggestion
**Theorem:** A non-empty convex polytope $$\mathcal{M}$$ can be characterized by a finite collection of linear inequality constraints, i.e.  
```

$$
\mathcal{M} = \left\{ \mu: a_j^T \mu \geq b_j, j \in \mathcal{J} \right\}, \qquad |\mathcal{J}| <\infty. 
$$

### Marginal Polytope

> Definition: For a distribution $$p(x)$$ and a sufficient statistics $$\phi(x)$$, the mean parameter is defined as:

Reviewer: ```suggestion
**Definition:** For a distribution $$p(x)$$ and a sufficient statistics $$\phi(x)$$, the mean parameter is defined as:
```
